VCUE-1187 add configuration managment interface

matteo-chesi opened the pull request:
The configuration management interface is composed by 3 functions:

- load_config - to load system configuration without specifying anything
- load_config_path - to load system configuration providing Some(path)
- update_config_by_user - to merge EDF annotations into system configuration to obtain the configuration specified by the user.

matteo-chesi committed:
VCUE-1187 - improve configuration management

matteo-chesi committed:
Add test for EDF annotation variable expansion when merging configuration

matteo-chesi committed:
pub adjustements

matteo-chesi committed:
cargo update

matteo-chesi committed:
Enable expanding with job environment

gwangmu committed:
Fix 'cargo test' compile error

gwangmu commented on src/config.rs:
```diff
@@ -49,13 +227,220 @@ fn load_raw_config(filepath: String) -> RawConfig {
+        *optstr =
+            Some(expand_vars_string(optstr.clone().unwrap(), env_option).unwrap_or(String::from("")));
+    }
+}
```
Should we raise an error if some environment variables are not bound (e.g., because of typos)? System configs are critical to functionality, so I think some pedantic sanity checks might be valuable as a precaution.

I think this can be done by adding `set -u` in the command (`-c`) of `expand_vars_string_with_env` and making `expand_raw_option_string` to return `SarusResult`. The former change will make all EDF variables also reject any unbound variables, but we could instruct users to use`:-` in  `${VAR:-}`.

matteo-chesi replied:
I agree, but how to do that ?
Just returning an error, we are invalidating the configuration load for something that maybe is not really important.
We cannot just print a warning, because we are in a library and we have no way to handle I/O.
We can return an error with the wrong expanded value inside together with a warning message, and the caller can decide what to do with it.

What do you think ?

gwangmu replied:
I think the last option sounds reasonable. I was just wondering what would be the best way to deliver the error message to the caller. Do you have any idea?

gwangmu commented on src/config.rs:
```diff
@@ -49,13 +227,220 @@ fn load_raw_config(filepath: String) -> RawConfig {
         }
     };
 
-    let r: RawConfig = toml_value;
+    let mut r: RawConfig = toml_value;
+
+    // Expand variables in the fields
+    expand_raw_config_fields(&mut r, env_option);
+
     r
 }
 
+fn expand_raw_config_fields(r: &mut RawConfig, e: &Option<HashMap<String,String>>) {
+    expand_raw_option_string(&mut r.edf_system_search_path, e);
+    expand_raw_option_string(&mut r.parallax_imagestore, e);
+    expand_raw_option_string(&mut r.parallax_mount_program, e);
+    expand_raw_option_string(&mut r.parallax_path, e);
+    expand_raw_option_string(&mut r.podman_module, e);
+    expand_raw_option_string(&mut r.podman_path, e);
+    expand_raw_option_string(&mut r.podman_tmp_path, e);
+    expand_raw_option_string(&mut r.runtime_path, e);
+    expand_raw_option_string(&mut r.tracking_tool, e);
+}
+
+fn expand_raw_option_string(optstr: &mut Option<String>, env_option: &Option<HashMap<String,String>>) {
+    if optstr.is_some() {
+        *optstr =
+            Some(expand_vars_string(optstr.clone().unwrap(), env_option).unwrap_or(String::from("")));
+    }
+}
+
 pub fn load_config() -> Config {
-    let config_file_path = String::from(CONFIG_FILE);
-    let r = load_raw_config(config_file_path);
+    load_config_path(None, &None)
+}
```
How do you think about changing the return value from `Config` to `SarusResult<Config>`? I think it's done for the EDF counterpart (`render` in src/lib.rs). My concern was that the caller had no way to determine whether config loading resulted in an error.

matteo-chesi replied:
As I wrote above, I agree but we must pay attention how we do that and what the caller can do with an error at this early stage.

gwangmu replied:
I had some thoughts, then I remembered some Go code, where the return value (the requested value) and the result (ok or not?) are returned separately, like here: https://github.com/google/syzkaller/blob/master/vm/vm.go#L196-L198.

So maybe we could restructure `SarusResult<T>` as `(T, Result<(), SarusError>)` ? Then the caller inspects the `Result` part to decide whether or not to drop `T`. Or maybe instead `Result<(T, Some(SarusError)), SarusError>`. This way, warnings can be attached to `T` if any.

I saw this pattern fairly frequently in Go, but I'm not sure how idiomatic it is in Rust, though. If you think it's a good idea, I can work on it separately after closing this MR.

gwangmu commented on src/config.rs:
```diff
@@ -49,13 +227,220 @@ fn load_raw_config(filepath: String) -> RawConfig {
+    load_config_path(None, &None)
+}
+
+pub fn load_config_path(config_option: Option<PathBuf>, env_option: &Option<HashMap<String,String>>) -> Config {
```
The comment about the return value (`Config` to `SarusResult<Config>`) also applies here.

matteo-chesi replied:
:+1:

gwangmu commented on src/config.rs:
```diff
@@ -49,13 +227,220 @@ fn load_raw_config(filepath: String) -> RawConfig {
+    rcfg
+}
+
+pub fn update_config_by_user(config: &mut Config, edf: EDF) {
```
It's a similar return value comment, but would it make sense to return `SarusResult<()>` here? Just to be more future-proof when it "may" return an error, even though there's nothing that can throw an error right now. I also think the public functions of a library may have to return some ACK to callers, because anything below them is opaque to the caller's view.

matteo-chesi replied:
:+1:

gwangmu committed:
Remove debugging code in a test

gwangmu reviewed (commented):
I have two additional comments that don't strictly correspond to code lines.

- About toml/config organization in the repository, how about moving `src/toml` to `test`, too? I think it makes more sense to move `src/toml` out of `src` than to move `config` into `src`.
- I think this MR doesn't include the support for unknown EDF parameters. I think we could either comment out src/lib.rs:406 (`validate_file(path, schema_content`) or auto-copy unknown EDF fields to `com.sarus.*` annotations. They can be restored or disabled before we release `raster` into the production system.

matteo-chesi commented:
> About toml/config organization in the repository, how about moving src/toml to test, too? I think it makes more sense to move src/toml out of src than to move config into src.

I agree, I can work on it.

gwangmu commented:
> I think this MR doesn't include the support for unknown EDF parameters. I think we could either comment out src/lib.rs:406 (validate_file(path, schema_content) or auto-copy unknown EDF fields to com.sarus.* annotations. They can be restored or disabled before we release raster into the production system.

On second thought, I think we probably need to do both because the "unknown" parameters should be visible to `Config` anyway.

matteo-chesi commented:
> I think this MR doesn't include the support for unknown EDF parameters. I think we could either comment out src/lib.rs:406 (validate_file(path, schema_content) or auto-copy unknown EDF fields to com.sarus.* annotations. They can be restored or disabled before we release raster into the production system.

No, it does include it.
I changed additionalProprerties to "true" in the schema at:
https://github.com/sarus-suite/raster/blob/VCUE-1187/src/schema/edf.json#L7
This way the toml can contain additional items without making it invalid.

Moreover I added a test here:
https://github.com/sarus-suite/raster/blob/VCUE-1187/src/lib.rs#L928-L934

To validate this otherwise invalid toml:
https://github.com/sarus-suite/raster/blob/VCUE-1187/src/toml/unknown_entry.toml

gwangmu commented:
Oh I'm sorry. I missed that part. I saw `validate_file` was intact, and thought unknown fields would generate errors there.

That said, I still have a question. Are unknown fields gonna be visible to `Config`? I'm concerned about the situation where we assume every configuration is passed to `Config` but some legacy unknown fields are not visible there.

matteo-chesi commented:
> Are unknown fields gonna be visible to `Config`? I'm concerned about the situation where we assume every configuration is passed to `Config` but some legacy unknown fields are not visible there.

No, they won't be visible, we are not parsing them, but I guess that's the agreed behavior, what can we do with entries that we do not know anything about ?
Environment variables and annotations are a different story, we don't know what can be listed there, but as long as they will be key/value strings they will be parsed.

gwangmu commented:
> No, they won't be visible, we are not parsing them, but I guess that's the agreed behavior, what can we do with entries that we do not know anything about ?

Off the top of my head, since this is just temporary for our development, maybe we could copy all unknown fields `XXX` to the annotation `com.sarus.XXX`. I'm assuming we're not gonna allow any unknown fields in the production release.

matteo-chesi committed:
multiple changes proposed on MR review

matteo-chesi commented:
> Off the top of my head, since this is just temporary for our development, maybe we could copy all unknown fields XXX to the annotation com.sarus.XXX. I'm assuming we're not gonna allow any unknown fields in the production release.

I don't know how to do that (parsing unknown entries) with current functions, and I don't understand the purpose of it.
As far as I understood from our last meeting conversation, allowing unknown entries is done just because we do not want that current EDFs, that work with pyxis, become suddenly invalid when passing to Skybox.

gwangmu commented:
> I don't know how to do that (parsing unknown entries) with current functions, and I don't understand the purpose of it.
As far as I understood from our last meeting conversation, allowing unknown entries is done just because we do not want that current EDFs, that work with pyxis, become suddenly invalid when passing to Skybox.

I also understand how we agreed, but I had a slightly different but related concern. Let me use our next meeting to check whether my concern is valid. I think we can skip this part in this MR.

matteo-chesi committed:
Add force_expand option to handle different use cases

matteo-chesi committed:
Fixed tests with force_expand

Madeeks commented:
I have integrated this into sarusctl with https://github.com/sarus-suite/sarusctl/compare/integrate_config.
Initial testing on Zinal is successful and quite promising so it would be great if we could proceed towards completing this MR.

gwangmu committed:
minor: move test/toml/plain.txt to test/etc/plain.txt

gwangmu committed:
minor: add a space after bash 'set -u'

gwangmu committed:
(Managed to) make 'toml_read' work

gwangmu committed:
Replace the TOML read logic to 'toml_read'

gwangmu committed:
cargo fmt

gwangmu committed:
Make 'force_expand' as an enum

gwangmu reviewed (approved):
I think it's a solid implementation. I made 'force_expand' an enum (VarExpand), but you could change the enum or keep it to the original version (by reverting the last commit).

matteo-chesi commented:
> I think it's a solid implementation. I made 'force_expand' an enum (VarExpand), but you could change the enum or keep it to the original version (by reverting the last commit).

Thank you Gwangmu, I am going to merge as it is.
